README.md
# Express JS Rocket powered middleware

- [x] ExpressJS server (on `PORT || 3000`)
- [x] Login and logout support
- [x] Google Login preconfigured
- [x] Local testing login available (test:123)
- [x] MongoDB sessions
- [x] MongoDB user
- [x] Overridable pages
- [x] JWT authentication


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.18](https://github.com/huksley/express-rocket/compare/v0.1.17...v0.1.18) (2021-01-03)


### Features

* overridable nav ([2ff48c4](https://github.com/huksley/express-rocket/commit/2ff48c4215cde002aab18cd2fb1c4ca3d94177e5))

### [0.1.17](https://github.com/huksley/express-rocket/compare/v0.1.16...v0.1.17) (2021-01-03)


### Features

* allow to generate login link for more scopes ([8e0d044](https://github.com/huksley/express-rocket/commit/8e0d044e68073513b57363e4347ab3ea02647605))

### [0.1.16](https://github.com/huksley/express-rocket/compare/v0.1.15...v0.1.16) (2021-01-03)


### Features

* store scope and provider in user profile ([c296c4f](https://github.com/huksley/express-rocket/commit/c296c4f3f0ddc3a9fe94fb7f326617641013a25b))

### [0.1.15](https://github.com/huksley/express-rocket/compare/v0.1.14...v0.1.15) (2021-01-02)


### Bug Fixes

* compilation error ([b31d9ad](https://github.com/huksley/express-rocket/commit/b31d9adc5b5207bcdb7750b0725ccc2c2fe42232))

### [0.1.14](https://github.com/huksley/express-rocket/compare/v0.1.13...v0.1.14) (2021-01-02)


### Bug Fixes

* Fixed access logging ([61bf3ca](https://github.com/huksley/express-rocket/commit/61bf3cac2adcda024fd730dfb2b447809d90e44e))

### [0.1.13](https://github.com/huksley/express-rocket/compare/v0.1.12...v0.1.13) (2021-01-01)


### Bug Fixes

* dao proper promise reject ([0ebff62](https://github.com/huksley/express-rocket/commit/0ebff62f5d90e544618aec2c8c7874650846a915))
* send message with json error http response ([d9faea5](https://github.com/huksley/express-rocket/commit/d9faea5930bac9a41bf97ceb1970a8f42d047fae))

### [0.1.12](https://github.com/huksley/express-rocket/compare/v0.1.11...v0.1.12) (2021-01-01)


### Bug Fixes

* properly save tokens ([3e8a8cd](https://github.com/huksley/express-rocket/commit/3e8a8cd118adfd6b8758a831ec70cb0424f7384f))

### [0.1.11](https://github.com/huksley/express-rocket/compare/v0.1.10...v0.1.11) (2021-01-01)


### Bug Fixes

* properly update profile on login ([8462fe3](https://github.com/huksley/express-rocket/commit/8462fe349ae86a18571430ee023fdcd40e85c6a6))

### [0.1.10](https://github.com/huksley/express-rocket/compare/v0.1.9...v0.1.10) (2021-01-01)


### Features

* app.findUser method ([4a7f5ef](https://github.com/huksley/express-rocket/commit/4a7f5efa797bf39a20edff8af062e58c63ad46e9))

### [0.1.9](https://github.com/huksley/express-rocket/compare/v0.1.8...v0.1.9) (2021-01-01)


### Bug Fixes

* livereload is prod dependency ([8ef81ba](https://github.com/huksley/express-rocket/commit/8ef81ba3689610efaaeb2f8d14f728164268d390))

### [0.1.8](https://github.com/huksley/express-rocket/compare/v0.1.7...v0.1.8) (2021-01-01)


### Features

* rename user page to dashboard ([286cfe2](https://github.com/huksley/express-rocket/commit/286cfe20146915483e0185ec57058abb21579b1f))

### [0.1.7](https://github.com/huksley/express-rocket/compare/v0.1.6...v0.1.7) (2020-12-27)


### Features

* head tags support ([c4eb2db](https://github.com/huksley/express-rocket/commit/c4eb2dbb7f77ff3da034faa4ff5cb0c0fb453f16))

### [0.1.6](https://github.com/huksley/express-rocket/compare/v0.1.5...v0.1.6) (2020-12-26)


### Features

* configurable google scopes ([289db2f](https://github.com/huksley/express-rocket/commit/289db2f21538db2f4b1db24a8701418813e1bee5))
* dynamic title and user page ([97d8712](https://github.com/huksley/express-rocket/commit/97d871210b970df9f9fc5cad50524e188aee3dad))
* known pages ([0015494](https://github.com/huksley/express-rocket/commit/0015494562a6090737a7aa4e862b12716a4a79b1))


### Bug Fixes

* bug fixes ([a7aefdb](https://github.com/huksley/express-rocket/commit/a7aefdb8b6251ab7fa4a69e5d0d42d22642feeb6))
* make react router work ([cba21fc](https://github.com/huksley/express-rocket/commit/cba21fcd930ad1469c8ee7ed08d14fa9310393e3))

### [0.1.5](https://github.com/huksley/express-rocket/compare/v0.1.4...v0.1.5) (2020-12-25)


### Features

* add dao and crud methods ([2d4ca46](https://github.com/huksley/express-rocket/commit/2d4ca460015016f56adc70fdf0412b3a630a7f54))

### [0.1.4](https://github.com/huksley/express-rocket/compare/v0.1.3...v0.1.4) (2020-12-07)


### Bug Fixes

* repo field in package.json ([4c6847b](https://github.com/huksley/express-rocket/commit/4c6847bc09430b6c437abceb20e4427327ef02e8))

### [0.1.3](https://github.com/huksley/express-rocket/compare/v0.1.2...v0.1.3) (2020-12-07)

### [0.1.2](https://github.com/huksley/express-rocket/compare/v0.1.1...v0.1.2) (2020-12-07)


### Features

* user html script example ([d63e842](https://github.com/huksley/express-rocket/commit/d63e8426d845cfeff33e9d9208076ff847b31a84))


### Bug Fixes

* fix serverless support ([5d65301](https://github.com/huksley/express-rocket/commit/5d65301f1a37fbbd4a0a01347636547263972ebd))
* proper user var for templates ([3682266](https://github.com/huksley/express-rocket/commit/368226614a9f0f8e7ff4e851918a653e57d82c0e))
* use PORT env var ([5f06f91](https://github.com/huksley/express-rocket/commit/5f06f91fc18b4c667d9c98b7e3101eaa9c8daf89))

### 0.1.1 (2020-12-06)


### Features

* initial release ([6811bf2](https://github.com/huksley/express-rocket/commit/6811bf2e9da570e725cac34ae7cafebb6bbe2380))


lib/dao.js
const logger = require('./logger')
const mongoose = require('mongoose')
const nanoid = require('nanoid')
const R = require('ramda')

let dao = null

class DAO {
  /**
   * Create DAO instance.
   *
   * @param {mongoose.Connection} conn - Connection to MongoDB
   */
  constructor(conn, options) {
    /**
     * Private connection property
     *
     * @type {mongoose.Connection}
     * @public
     */
    this.connection = conn
    this.options = options
  }

  model(name, schema) {
    const model = (this.options && this.options.modelPrefix ? this.options.modelPrefix : '') + name
    const existing = this.connection.models[model]
    if (existing) {
      logger.verbose('Returning existing model', model)
      return existing
    } else {
      return this.connection.model(model, schema)
    }
  }

  close() {
    logger.verbose('Closing MongoDB connection')
    this.connection.close()
    this.connection = null
    dao = null
  }

  generateId() {
    return nanoid.nanoid(16)
  }
}

const create = function (options) {
  return new Promise(async resolve => {
    const url = process.env.MONGO_URL
    const mongoUrl = url
      .replace('$MONGO_PASSWORD', process.env.MONGO_PASSWORD)
      .replace('$MONGO_USER', process.env.MONGO_USER)

    if (dao === null) {
      logger.verbose('Connecting to ' + url)
      const conn = await mongoose.createConnection(mongoUrl, {
        bufferCommands: false,
        bufferMaxEntries: 0,
        useNewUrlParser: true,
        useUnifiedTopology: true,
        useCreateIndex: true,
        useFindAndModify: false
      })

      dao = new DAO(conn, options)
      logger.verbose('Created dao', Object.keys(dao))
    }

    resolve(dao)
  })
}

/**
 * Initializes data access APIs and models.
 *
 * @returns {Promise<DAO>} DAO with models ready to use
 */
module.exports = {
  create,
  define: function (name, model, properties, options) {
    options = options || {}
    const schema = new mongoose.Schema({
      _id: {
        type: String,
        default: () => nanoid.nanoid(16)
      },
      updatedAt: Date,
      ...model
    })

    return {
      schema,
      name,
      properties,
      json: obj => (obj === undefined ? obj : JSON.parse(JSON.stringify(obj))),
      // Single record
      find: filter =>
        new Promise((resolve, reject) => {
          create()
            .then(dao => {
              const Model = dao.connection.model(name, schema)
              Model.find(filter)
                .exec()
                .then(records => {
                  if (records.length == 0) {
                    resolve(undefined)
                  } else if (records.length > 1) {
                    logger.warn('Consistency error, found ' + name + ' records: ', records.length)
                    resolve(R.head(records))
                  } else {
                    resolve(R.head(records))
                  }
                })
                .catch(reject)
            })
            .catch(reject)
        }),
      delete: id => {
        return new Promise((resolve, reject) =>
          create()
            .then(dao => {
              const Campaign = dao.model(name, schema)
              Campaign.find({ _id: id })
                .exec()
                .then(records => {
                  const existing = R.head(records)
                  if (!existing) {
                    logger.warn('Specified id cannot be found: ', id)
                    resolve(undefined)
                  } else {
                    logger.verbose('Deleting', id)
                    existing.deleteOne()
                    resolve(existing)
                  }
                })
                .catch(reject)
            })
            .catch(reject)
        )
      },
      // Filter all records
      filter: filter =>
        new Promise((resolve, reject) => {
          create(options)
            .then(dao => {
              const Model = dao.connection.model(name, schema)
              Model.find({ ...filter })
                .exec()
                .then(records => {
                  resolve(records)
                })
                .catch(reject)
            })
            .catch(reject)
        }),
      create: obj => {
        return new Promise((resolve, reject) =>
          create()
            .then(dao => {
              const Model = dao.model(name, schema)
              logger.verbose('Creating', obj)
              new Model({
                ...obj,
                updatedAt: new Date()
              })
                .save()
                .then(resolve)
                .catch(reject)
            })
            .catch(reject)
        )
      },
      // Add or update one based on filter
      upsert: (filter, obj) => {
        const filtered = R.pick(properties, obj)
        return new Promise((resolve, reject) =>
          create()
            .then(dao => {
              const Model = dao.model(name, schema)
              Model.find({ ...filter })
                .exec()
                .then(records => {
                  if (records.length == 0) {
                    logger.verbose('Creating', obj)
                    new Model({
                      ...obj,
                      updatedAt: new Date()
                    })
                      .save()
                      .then(resolve)
                      .catch(reject)
                  } else {
                    if (records.length > 1) {
                      logger.warn('Consistency error, found records: ', records.length)
                    }
                    const existing = R.head(records)
                    const existingFiltered = R.pick(properties, existing)
                    if (!R.equals(filtered, existingFiltered)) {
                      logger.verbose('Updating record')
                      options.preUpdate && options.preUpdate(existing, obj)
                      existing
                        .set({
                          ...obj,
                          updatedAt: new Date()
                        })
                        .save()
                        .then(resolve)
                        .catch(reject)
                    } else {
                      logger.verbose('Record not changed')
                      resolve(existing)
                    }
                  }
                })
                .catch(reject)
            })
            .catch(reject)
        )
      }
    }
  }
}


lib/user.js
const logger = require('./logger')
const mongoose = require('mongoose')
const nanoid = require('nanoid')
const dao = require('./dao')
const R = require('ramda')

/**
 * @class UserSchema
 */
const UserSchema = new mongoose.Schema({
  _id: {
    type: String,
    default: () => nanoid.nanoid(16)
  },
  email: { type: String, index: true, unique: true },
  token: String,
  displayName: String,
  credentials: mongoose.Schema.Types.Mixed,
  provider: String,
  scope: String,
  updatedAt: Date
})

const UserSchemaProperties = ['email', 'token', 'displayName', 'credentials', 'provider', 'scope']

const User = {
  find: id =>
    new Promise((resolve, reject) => {
      dao
        .create(options)
        .then(dao => {
          const User = dao.connection.model('User', UserSchema)
          User.find({ _id: id })
            .exec()
            .then(users => {
              if (users.length == 0) {
                resolve(undefined)
              } else if (users.length > 1) {
                logger.warn('Consistency error, found users: ', users.length)
                resolve(R.head(users).toObject())
              } else {
                resolve(R.head(users).toObject())
              }
            })
            .catch(reject)
        })
        .catch(reject)
    }),
  upsert: (filter, user) => {
    user = R.pick(UserSchemaProperties, user)
    return new Promise((resolve, reject) =>
      dao
        .create()
        .then(dao => {
          const User = dao.model('user', UserSchema)
          User.find(filter)
            .exec()
            .then(users => {
              if (users.length === 0) {
                logger.verbose('Creating user', user)
                new User({
                  ...user,
                  updatedAt: new Date()
                })
                  .save()
                  .then(resolve)
                  .catch(reject)
              } else {
                if (users.length > 1) {
                  logger.warn('Consistency error, found users: ', users.length)
                }
                const existing = R.head(users)
                if (!R.equals(user, R.pick(UserSchemaProperties, existing))) {
                  user.credentials = Object.assign(existing.credentials || {}, user.credentials || {})
                  existing.attributes = user.attributes
                  existing.markModified('attributes')

                  logger.verbose(
                    'Updating user',
                    user,
                    'existing',
                    R.pick(UserSchemaProperties, existing.toObject())
                  )
                  existing
                    .set({
                      ...user,
                      updatedAt: new Date()
                    })
                    .save()
                    .then(resolve)
                    .catch(reject)
                } else {
                  logger.verbose('User not changed', user.email)
                  resolve(existing)
                }
              }
            })
            .catch(reject)
        })
        .catch(reject)
    )
  }
}

module.exports = {
  User
}


lib/token.js
const jwt = require('jsonwebtoken')
const logger = require('./logger')

// Get auth header value, and if successful, attach token data to request
const checkTokenMiddleware = async (req, res, next) => {
  const bearerHeader = req.headers['authorization']
  if (typeof bearerHeader !== 'undefined') {
    const token = bearerHeader.split(' ')[1]
    if (token === undefined || token === '') {
      next({ statusCode: 403, message: 'No token specified' })
    }

    verifyToken(token)
      .then(data => {
        logger.verbose('Authenticated with token', data)
        req.auth = data
        next()
      })
      .catch(error => {
        logger.warn('Authentication with token failed', error)
        next(error)
      })
  } else {
    next({ statusCode: 403, message: 'Not authenticated, no token found' })
  }
}

// Verify token validity
const verifyToken = token => {
  return new Promise((resolve, reject) => {
    jwt.verify(token, process.env.SESSION_SECRET, (err, data) => {
      if (err) {
        reject(err)
      } else {
        resolve(data)
      }
    })
  })
}

// Issue Token
const signToken = data => {
  return new Promise((resolve, reject) => {
    jwt.sign(data, process.env.SESSION_SECRET, { expiresIn: '365 day' }, (err, token) => {
      if (err) {
        reject(err)
      } else {
        logger.verbose('Created token', token)
        return resolve(token)
      }
    })
  })
}

module.exports.middleware = checkTokenMiddleware
module.exports.signToken = signToken
module.exports.verifyToken = verifyToken


lib/logger.js
/* eslint-disable no-console */

const env = process != undefined && process.env !== undefined ? process.env : {}

// @ts-ignore
if (!console.__logging_enabled) {
  // @ts-ignore
  console.__logging_enabled = true

  const _info = console.info
  const _error = console.error
  const _warn = console.warn
  const _prefix = env.GIT_TAGS || 'dev'

  console.info = (...args) => _info.apply(console, [_prefix, ...args])
  console.error = (...args) => _error.apply(console, [_prefix, ...args])
  console.warn = (...args) => _warn.apply(console, [_prefix, ...args])
}

const fs =
  typeof window === 'object'
    ? {
        writeFileSync: (name, contents) => {
          console.log(name, contents)
        }
      }
    : eval('require("fs")')

module.exports = {
  isAudit: !!env.LOG_AUDIT,
  isVerbose: !!env.LOG_VERBOSE,
  isDump: !!env.LOG_DUMP,
  info: console.info,
  audit: !!env.LOG_AUDIT ? (...args) => console.info.apply(console, ['AUDIT', ...args]) : () => {},
  verbose: !!env.LOG_VERBOSE ? (...args) => console.info.apply(console, ['DEBUG', ...args]) : () => {},
  dump: (filename, contents) => fs.writeFileSync(filename, contents),
  warn: console.warn,
  error: console.error
}


index.js
const logger = require('./lib/logger')
const express = require('express')
const cors = require('cors')
const helmet = require('helmet')
const app = express()
const path = require('path')
const port = process.env.PORT || 3000
const bodyParser = require('body-parser')
const serverless = require('serverless-http')
const cookieParser = require('cookie-parser')
const nocache = require('nocache')
const session = require('express-session')
const passport = require('passport')
const GoogleStrategy = require('passport-google-oauth20').Strategy
const { middleware: checkToken, signToken } = require('./lib/token')
const MongoDBStore = require('connect-mongodb-session')(session)
const LocalStrategy = require('passport-local').Strategy
const ejs = require('ejs')
const { User } = require('./lib/user')
const dao = require('./lib/dao')
const fetch = require('node-fetch')

app.set('view engine', 'ejs')
app.set('views', path.join(__dirname, 'views'))
ejs.rmWhitespace = true
ejs.openDelimiter = '{'
ejs.closeDelimiter = '}'

const launch = options => {
  options = options ? options : {}
  const content = options.content ? options.content : {}
  options.content = content
  app.options = options

  const url = process.env.MONGO_URL
  if (!url) {
    throw new Error('MONGO_URL not defined')
  }
  const mongoUrl = url
    .replace('$MONGO_PASSWORD', process.env.MONGO_PASSWORD)
    .replace('$MONGO_USER', process.env.MONGO_USER)

  var store = new MongoDBStore({
    uri: mongoUrl,
    collection: (options.modelPrefix || 'Rocket') + 'Sessions',
    expires: 1000 * 60 * 60 * 24 * 30, // 30 days in milliseconds
    connectionOptions: {
      bufferMaxEntries: 0,
      useNewUrlParser: true,
      useUnifiedTopology: true,
      serverSelectionTimeoutMS: 10000
    }
  })

  app.use(
    session({
      cookie: {
        maxAge: 1000 * 60 * 60 * 24 * 30, // 30 days in milliseconds
        secure: process.env.NODE_ENV === 'production',
        httpOnly: true
      },
      resave: false,
      saveUninitialized: false,
      secret: process.env.SESSION_SECRET,
      store: store
    })
  )

  app.use(
    helmet.hsts({
      maxAge: 31536000,
      includeSubDomains: false
    })
  )

  app.use(
    helmet.referrerPolicy({
      policy: 'origin-when-cross-origin'
    })
  )

  app.use(passport.initialize())
  app.use(passport.session())
  app.locals.livereload = !!global.livereload

  const baseUrl = process.env.BASE_URL || `http://localhost:${port}`
  const allowlist = process.env.BASE_URL
    ? [`http://localhost:${port}`, process.env.BASE_URL]
    : [`http://localhost:${port}`]
  const STRATEGY = process.env.PASSPORT_STRATEGY || 'local'

  const config = {
    clientID: process.env.GOOGLE_CLIENT_ID,
    clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    callbackURL: baseUrl + '/auth/google/callback'
  }

  // Include request parsers
  app.use(bodyParser.json())
  app.use(bodyParser.urlencoded({ extended: false }))
  app.use(cookieParser())
  app.disable('etag')

  app.use(
    cors((req, callback) => {
      if (allowlist.indexOf(req.header('Origin')) !== -1) {
        callback(null, { origin: true })
      } else {
        if (req.header('Origin') !== undefined) {
          logger.warn('Blocked origin', req.header('Origin'))
        }
        callback(null, { origin: false })
      }
    })
  )

  // Disable caching
  app.use(nocache())

  // Log requests and responses
  if (logger.isVerbose) {
    const staticMatch = /.*\.(svg|png|gif|jpg|js|css|ico|json)$/i
    app.use((request, response, next) => {
      const { method, url } = request
      const ignore = staticMatch.test(url)
      if (ignore) {
        next()
      } else {
        const end = response.end
        const started = new Date().getTime()
        logger.verbose(`--> ${method} ${url}`)
        response.end = (...args) => {
          const { statusCode } = response
          const now = new Date().getTime()
          const elapsed = now - started
          logger.verbose(`<-- ${statusCode} ${method} ${url} Δ ${elapsed}ms`)
          end.apply(null, args)
        }
        next()
      }
    })
  }

  if (options.staticPath) {
    app.use(express.static(options.staticPath))
  }

  // Mount the static files directory
  app.use(express.static(path.join(__dirname, 'public')))

  app.get('/', async (req, res) => {
    res.render('index', {
      baseUrl,
      session: req.session,
      user:
        req.session.passport && req.session.passport.user && req.session.passport.user.token
          ? req.session.passport.user
          : undefined,
      title: content.title ? content.title(req, res) : 'Example',
      content: content.index ? await content.index(req, res) : undefined,
      options
    })
  })

  const knownPage = /^faq|changelog|privacy|tos|feedback|support|sandbox|pricing|integrations$/
  app.get('/:page', async (req, res, next) => {
    const page = req.params.page
    if (knownPage.test(page)) {
      const contentFunc = content[page]
      res.render('index', {
        baseUrl,
        session: req.session,
        user:
          req.session.passport && req.session.passport.user && req.session.passport.user.token
            ? req.session.passport.user
            : undefined,
        title: content.title ? content.title(req, res) : page,
        content: contentFunc ? await contentFunc(req, res) : '<h1>' + page + '</h1> No such page: ' + page,
        options
      })
    } else {
      next()
    }
  })

  app.get('/dashboard(/*)?', async (req, res) => {
    res.render('dashboard', {
      baseUrl,
      session: req.session,
      user:
        req.session.passport && req.session.passport.user && req.session.passport.user.token
          ? req.session.passport.user
          : undefined,
      title: content.title ? content.title(req, res) : 'User',
      content: content.dashboard ? await content.dashboard(req, res) : undefined,
      options
    })
  })

  app.checkToken = checkToken
  app.dao = dao.define
  app.checkAuth = (req, res, next) => {
    if (req.session && req.session.passport && req.session.passport.user) {
      next()
      return
    }

    checkToken(req, res, next)
  }
  app.findUser = User.find

  app.crud = (prefix, dao, options) => {
    options = options || {}
    const mw = options.middleware ? options.middleware : (_req, _res, next) => next()
    app.get(prefix, mw, async (req, res) => {
      const records = await dao.filter({ ...(options.filter && options.filter(req)) })
      logger.verbose('Listing', dao.name)
      res.status(200).json(records.map(dao.json))
    })

    app.post(prefix, mw, async (req, res) => {
      const { body } = req
      const obj = { ...body, ...(options.required && options.required(req)) }
      logger.verbose('Adding', obj)
      const created = await dao.create(obj)
      res.status(200).json({ ...dao.json(created) })
    })

    app.get(prefix + '/:id', mw, async (req, res) => {
      const id = req.params.id
      const filter = { _id: id, ...(options.required && options.required(req)) }
      logger.verbose('Getting', filter)
      const obj = await dao.find(filter)
      if (obj === undefined) {
        res.status(404).json({ message: 'Not found' })
      } else {
        res.status(200).json({ ...dao.json(obj) })
      }
    })

    app.patch(prefix + '/:id', mw, async (req, res) => {
      const id = req.params.id
      const { body } = req
      const filter = { _id: id, ...(options.required && options.required(req)) }
      logger.verbose('Updating', filter)
      const obj = await dao.find(filter)
      if (obj === undefined) {
        res.status(404).json({ message: 'Not found' })
      } else {
        const obj = await dao.upsert(filter, { ...filter, ...body })
        res.status(200).json({ ...dao.json(obj) })
      }
    })

    app.delete(prefix + '/:id', mw, async (req, res) => {
      const id = req.params.id
      const filter = { _id: id, ...(options.required && options.required(req)) }
      logger.verbose('Deleting', filter)
      const obj = await dao.find(filter)
      if (obj === undefined) {
        res.status(404).json({ message: 'Not found' })
      } else {
        await dao.delete(id)
        res.status(204).send(null)
      }
    })
  }

  app.userCrud = (prefix, dao, options) => {
    const updatedOptions = {
      middleware: app.checkAuth,
      filter: req => ({ userId: req.auth ? req.auth.userId : req.session.passport.user.userId }),
      required: req => ({
        userId: req.auth ? req.auth.userId : req.session.passport.user.userId
      }),
      ...options
    }
    app.crud(prefix, dao, updatedOptions)
  }

  app.get('/error', async (req, res) => {
    res.render('error', {
      baseUrl,
      session: req.session,
      user:
        session.passport && session.passport.user && session.passport.user.token
          ? session.passport.user
          : undefined,
      title: content.title ? content.title(req, res) : 'Error',
      content: content.error ? await content.error(req, res) : undefined,
      error: 'Unknown error occured'
    })
  })

  app.get('/auth/logout', function (req, res) {
    req.session.destroy()
    req.logout()
    res.redirect('/')
  })

  passport.serializeUser(function (user, cb) {
    cb(null, user)
  })

  passport.deserializeUser(function (obj, cb) {
    cb(null, obj)
  })

  if (STRATEGY === 'local') {
    passport.use(
      new LocalStrategy(function (username, password, done) {
        logger.verbose('Login username', username, 'password', password)
        const email = 'test@email.com'
        if (username === 'test' && password === '123') {
          const profile = {
            id: username,
            email,
            email_verified: true,
            displayName: 'Tester',
            provider: 'test',
            emails: [
              {
                value: email,
                email_verified: true
              }
            ]
          }
          const accessToken = process.env.LOCAL_GOOGLE_ACCESS_TOKEN || 'deadbeef'
          const refreshToken = process.env.LOCAL_GOOGLE_REFRESH_TOKEN || 'beefdead'
          profile.credentials = {
            accessToken,
            refreshToken
          }
          profile.email = email
          profile.scope = process.env.LOCAL_GOOGLE_SCOPE || 'email profile'
          User.upsert({ email }, profile).then(user => {
            signToken({
              id: profile.id,
              sub: profile.email,
              accessToken,
              refreshToken,
              userId: user._id
            }).then(token => {
              profile.token = token
              profile.userId = user._id
              done(null, profile)
            })
          })
        } else {
          return done(null, false)
        }
      })
    )

    app.get('/auth/login', (req, res) =>
      res.render('local-login', {
        baseUrl,
        session: req.session,
        title: 'Login',
        user:
          req.session.passport && req.session.passport.user && req.session.passport.user.token
            ? req.session.passport.user
            : undefined,
        options
      })
    )

    app.post(
      '/auth/submit',
      passport.authenticate('local', { failureRedirect: baseUrl + '/error' }),
      function (req, res) {
        if (options.onLogin) {
          options.onLogin(req, res)
        } else {
          res.redirect('/dashboard')
        }
      }
    )
  }

  if (STRATEGY === 'google') {
    passport.use(
      new GoogleStrategy(
        {
          ...config,
          passReqToCallback: true
        },
        (req, accessToken, refreshToken, profile, done) => {
          fetch(
            'https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=' + encodeURIComponent(accessToken)
          )
            .then(response => response.json())
            .then(tokenInfo => {
              const email = profile.email ? profile.email : profile.emails[0].value

              if (tokenInfo.error) {
                logger.warn('Failed to fetch token information', email, tokenInfo.error)
                return done(null, false)
              }

              logger.info('Logged in to Google', profile.id, email, tokenInfo)
              profile.credentials = {
                accessToken,
                refreshToken
              }
              profile.email = email
              profile.scope = tokenInfo.scope
              User.upsert({ email }, profile).then(user => {
                signToken({
                  id: profile.id,
                  sub: email,
                  accessToken,
                  refreshToken,
                  userId: user._id
                }).then(token => {
                  profile.token = token
                  profile.userId = user._id
                  done(null, profile)
                })
              })
            })
            .catch(error => {
              logger.warn('Failed to fetch token information', profile.email, error)
              return done(null, false)
            })
        }
      )
    )

    app.authHandler = scope =>
      passport.authenticate('google', {
        // https://developers.google.com/identity/protocols/oauth2/web-server
        // https://developers.google.com/identity/protocols/oauth2/openid-connect#scope-param
        scope: scope
          ? scope
          : process.env.GOOGLE_SCOPES
          ? process.env.GOOGLE_SCOPES.split(' ')
          : ['profile', 'email'],
        prompt: process.env.GOOGLE_PROMPT || 'consent',
        accessType: 'offline',
        includeGrantedScopes: true
      })

    app.get('/auth/login', app.authHandler())

    app.get(
      '/auth/google/callback',
      passport.authenticate('google', {
        failureRedirect: baseUrl + '/error'
      }),
      function (req, res) {
        if (options.onLogin) {
          options.onLogin(req, res)
        } else {
          res.redirect('/dashboard')
        }
      }
    )
  }

  if (options.apis) {
    options.apis(app, options)
  }

  app.use(function (err, req, res, _next) {
    const { statusCode, code, message } = err
    const httpCode = parseInt(statusCode || code || 500, 10)
    logger.warn('Error', err)
    if (req.headers.accept && req.headers.accept.indexOf('text/html') >= 0) {
      res
        .status(!isNaN(httpCode) && httpCode > 399 && httpCode < 599 ? httpCode : 500)
        .send(
          (message || 'Internal server error') +
            (global.livereload
              ? '<script async defer src="http://localhost:35729/livereload.js"></script>'
              : '')
        )
    } else {
      res
        .status(!isNaN(httpCode) && httpCode > 399 && httpCode < 599 ? httpCode : 500)
        .send({ message: message || 'Internal server error' })
    }
  })

  app.get('*', function (req, res) {
    if (req.headers.accept && req.headers.accept.indexOf('text/html') >= 0) {
      res
        .status(404)
        .send(
          'Not found' +
            (global.livereload
              ? '<script async defer src="http://localhost:35729/livereload.js"></script>'
              : '')
        )
    } else {
      res.status(404).send({ message: 'Not found' })
    }
  })

  process.on('beforeExit', code => {
    logger.verbose('NodeJS exiting', code)
  })

  process.on('SIGINT', signal => {
    logger.verbose('Caught interrupt signal', signal)
    process.exit(1)
  })

  // Do something when AWS lambda started
  if (process.env.AWS_EXECUTION_ENV !== undefined) {
    // _HANDLER contains specific invocation handler for this NodeJS instance
    logger.verbose('AWS Lambda started, handler:', process.env._HANDLER)
  } else {
    app.listen(port, () => logger.info(`API Server listening on port ${port}`))
  }

  return app
}

const serverlessHandler = app =>
  serverless(app, {
    binary: ['image/png', 'image/jpeg', 'image/x-icon']
  })

module.exports.serverless = app => async (event, context) => {
  context.callbackWaitsForEmptyEventLoop = false
  const result = serverlessHandler(app)(event, context)
  return result
}

module.exports.launch = launch
